include/FanFicArchive.py

```diff
@@ -0,0 +1,85 @@
+from bs4 import BeautifulSoup; from urllib import request; from re import match; import requests
+
+parser = "lxml"
+
+def fanfic_req(link): return BeautifulSoup(request.urlopen(link).read(), parser)
+def fanfic_check(id, name):
+    if isinstance(id, int): id = str(id)
+    else: assert isinstance(id, str), "Error for" + name + ": id not int or str"
+
+def ao3_rip(ao3_id): # downloads whole story in one go
+    fanfic_check(ao3_id, "ao3_rip")
+    text = fanfic_req('http://archiveofourown.org/works/'+ao3_id+'?view_adult=true&view_full_work=true')
+    tag_bar = text.find("dl", class_="stats").find_all("dd") # everything in the top bar
+    tag_list = {}
+    for i in tag_bar:
+        tag = [j.get_text() for j in i.find_all("a", class_="tag")]
+        if tag == []: tag = [i.get_text().lstrip('\n').lstrip(' ').rstrip(' ').rstrip('\n')]
+        tag_list[i.get("class")[0]] = tag
+    tag_list['author'] = [i.get_text() for i in text.find_all("a", rel="author")]
+    summary = text.find_all("div", class_="summary module", role="complementary") # just below top bar
+    chapter = {}
+    for i in text.find_all("div", class_="chapter", id=True): # separating each chapters
+        x = i.find("h3", class_="title").find("a").get("href").split("/")[-1]
+        chapter[i.get('id')] = [x, i]
+    return ao3_id, tag_list, summary, chapter
+
+def ff_rip(ff_id): # no image
+    fanfic_check(ff_id, "ff_rip")
+    text = fanfic_req("https://www.fanfiction.net/s/" +ff_id +"/1") # first chapter
+    tag_line = text.find("span", class_="xgray xcontrast_txt").get_text().split(" - ") # top bar
+    tag_list = {}
+    for i in tag_line:
+        if ":" in i: bracket, item = i.split(": ")[0:2]; tag_list[bracket] = item.rstrip(' ')
+        else: tag_list[['Language', 'Genre', 'Characters'][tag_line.index(i)-1]] = i
+    tag_list['Characters'] = tag_list['Characters'].split(', ')
+    tag_list['Genre'] = tag_list['Genre'].split('/')
+    del tag_list['id']
+    author = text.find("div", id="profile_top").find("a", class_="xcontrast_txt", title=False, target=False) # tag filtering
+    userid, username = author.get("href").split("/")[-2:] # the text in the tag, unlike the last line
+    summary = text.find("div", class_="xcontrast_txt", style=True).get_text() # tag filtering
+    chapter_list = {}
+    for i in range(1, int(tag_list['Chapters'])+1): # loop through each chapter
+        r = request.urlopen("https://www.fanfiction.net/s/"+ff_id+"/"+str(i)).read()
+        storyid = match(".*storytextid=([0-9]+)", str(r)).group(1) # can't get it any other way
+        text = BeautifulSoup(r, "lxml")
+        story = text.find("div", class_="storytext xcontrast_txt nocopy", id="storytext") # literal lines
+        chapter_list[i] = [storyid, story]
+    return ff_id, userid, username, summary, tag_list, chapter_list
```

Comment:
This tuple is heterogeneous relative to the one returned by the function above, and it looks like the others all are too. This suggests you want a class here to name the fields correctly.

Reply:
Might need to explain this one

Reply:
If you have some client code that deals with one function, you can't just swap it out for one of the other site scraping functions here, as the return types are incompatible. You could use a class with some common field names to allow some level of abstraction between them.